File: src/Parser.jl
using PyCall
const pysces = pyimport("pysces") #import pysces
using Gillespie
using Gadfly
import Random: seed!

model = pysces.model("DecayingDimerizing", dir="c:\\Pysces\\psc") #PyObject

# a `Vector` of `Int64`, representing the initial states of the system. i.e. states of S1, S2, S3. #InitVar
function getInitVar(model)
    species = model.species
    x1 = Vector{Int64}([])
    for i in species
        append!(x1, convert(Int64, eval(Meta.parse("model." * i))))
    end
    return x1
end

# a `Vector` of `Float64` representing the parameters of the system. #InitPar
function getInitParam(model)
    parameters = model.parameters
    x1 = Vector{Float64}([])
    for i in parameters
        append!(x1, eval(Meta.parse("model." * i)))
    end
    return x1
end

# Stoichiometric matrix
# get the Nmatrix using pysces, convert it to an array, then to an integer array, then find its transpose
function getMatrix(model)
    return (convert(Matrix{Int64}, model.Nmatrix.array))'
end

# gets the propensities by parsing through the pysces model and looking for the reactions
# returns a string that represents an array of propensities
function getPropensities()

    reactions = model.reactions
    output = "["
    
    open(model.ModelDir * "/" * model.ModelFile) do f

        current_reaction = 1
    
        # read till end of file
        while (! eof(f))
            line = readline(f) # read a new / next line for every iteration
            if (occursin(reactions[current_reaction], line))
                readline(f)
                output = output * readline(f) * ","
                if (reactions[current_reaction] == last(reactions))
                    break
                end
                current_reaction += 1
            end
        end
    end
    return replace(first(output, length(output) - 1) * "]\n"," " => "") # remove last comma, add closing bracket, remove spaces
end

# Constructs expression for propensities function
macro getF_dd()
    species = model.species
    parameters = model.parameters

    funcString = "function F_dd(x,parms)\n"
    stateString = "    ("
    paramString = "    ("
    

    for i in species
        stateString = stateString * i * ","
    end
    stateString = first(stateString, length(stateString) - 1) * ") = x\n"

    for i in parameters
        paramString = paramString * i * ","
    end
    paramString = first(paramString, length(paramString) - 1) * ") = parms\n"

    propString = getPropensities()

    return Meta.parse(funcString * stateString * paramString * "    " * propString * "end")
end

@getF_dd

function simulate()
    x0 = getInitVar(model) # a `Vector` of `Int64`, representing the initial states of the system. i.e. states of S1, S2, S3. #InitVar
    nu = getMatrix(model) # a `Matrix` of `Int64`, representing the transitions of the system, organised by row. Stoichiometric Matrix
    parms = getInitParam(model) # a `Vector` of `Float64` representing the parameters of the system. #InitPar

    # the final simulation time (`Float64`)
    tf = 10.0
    seed!(1234)

    result = ssa(x0,F_dd,nu,parms,tf)
    data = ssa_data(result)
end

simulate()

File: src/Parser.py
import sys
import pysces
import numpy as np

#python3 Parser.py DecayingDimerizing.psc

#name = sys.argv[1]
#name = "DecayingDimerizing"

# a `Vector` of `Int64`, representing the initial states of the system. i.e. states of S1, S2, S3. #InitVar
def get_initvar(model):
    output = "x0 = ["

    for i in model.species:
        output += (str(int(getattr(model, i))) + ",")

    output = output[0:-1] + "]\n"
    return output

def get_initparam(model):
    output = "parms = ["

    for i in model.parameters:
        output += (str(getattr(model, i)) + ",")

    output = output[0:-1] + "]\n"
    return output

def get_matrix(model):
    output = "nu = [["
    m = np.asarray(np.transpose(model.Nmatrix.array), dtype='int')
    for i in m:# removes "".psc" if in string

        for j in i:
            output += str(j) + " "
        output = output[0:-1] + "];["
    output = output[0:-2] + "]\n"
    return output

def get_propensities(model):
    reactions = model.reactions
    output = "["
    current_reaction = 0

    f = open(model.ModelDir + "/" + model.ModelFile, "r")

    for x in f:
        if reactions[current_reaction] in x:
            current_reaction += 1
            f.readline()
            output += str(f.readline().split())[2:-2] + ","
            if current_reaction >= len(reactions):
                break
    f.close()
    return output[0:-1] + "]\n"

def get_F_dd(model):
    funcString = "function F_dd(x,parms)\n"
    stateString = "    ("
    paramString = "    ("
    popString = "    " + get_propensities(model)

    for i in model.species:
        stateString += str(i) + ","
    stateString = stateString[0:-1] + ") = x\n"

    for i in model.parameters:
        paramString += str(i) + ","
    paramString = paramString[0:-1] + ") = parms\n"

    return(funcString + stateString + paramString + popString + "end\n")

def write_file(name):

    if (name.find(".psc") != -1):
        name = name[0:name.find(".psc")]

    model = pysces.model(name + ".psc", dir='c:\\Pysces\\psc') #create pysces model
    output = "using Gillespie\nimport Random: seed!\n" + get_F_dd(model) + get_initvar(model) + get_matrix(model) + get_initparam(model) + "tf = 10.0\nseed!(1234)\nresult = ssa(x0,F_dd,nu,parms,tf)"
    f = open(name + ".jl", "w") #Create new Julia file
    f.write(output)
    f.close
    
write_file("DecayingDimerizing")